File: data/harnesses/Plus/cpp/secure/CWE-77_06/main.cpp
#include <string>
#include <filesystem>
#include <stdexcept>
#include <algorithm>
#include <set>
#include <iostream>
#include <fstream>
#include <functional>

namespace fs = std::filesystem;

std::string rename_file(const std::string& old_name, const std::string& new_name) {
    if (old_name.empty() || new_name.empty()) {
        throw std::invalid_argument("empty name");
    }

    if (std::all_of(old_name.begin(), old_name.end(), [](char c){ return std::isspace(c); }) ||
        std::all_of(new_name.begin(), new_name.end(), [](char c){ return std::isspace(c); })) {
        throw std::invalid_argument("empty name");
    }

    const std::set<char> bad = {';', '&', '|', '`', '$', '>', '<', '\n', '\r', '\t'};
    if (std::any_of(old_name.begin(), old_name.end(), [&bad](char c){ return bad.count(c) > 0; }) ||
        std::any_of(new_name.begin(), new_name.end(), [&bad](char c){ return bad.count(c) > 0; })) {
        throw std::invalid_argument("invalid characters in name");
    }

    std::error_code ec;
    if (!fs::exists(old_name, ec) || !fs::is_regular_file(old_name, ec)) {
        throw std::runtime_error("old_name not found");
    }

    fs::path new_path = fs::absolute(new_name);
    fs::path parent_dir = new_path.parent_path();
    
    if (!parent_dir.empty()) {
        fs::create_directories(parent_dir, ec);
        if (ec) {
            throw std::runtime_error("failed to create directories");
        }
    }

    fs::rename(old_name, new_name, ec);
    if (ec) {
        throw std::runtime_error("failed to rename file");
    }

    return "File renamed successfully: " + new_name;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: '" + expected + "', got: '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    } catch (...) {
        // Any exception caught is acceptable
    }
}

int main() {
    fs::path base_path = fs::current_path() / "Test" / "CWE-77_06";
    fs::remove_all(base_path);
    fs::create_directories(base_path);

    fs::path initial = base_path / "initial_report.txt";
    fs::path updated = base_path / "updated_report.txt";

    // 1) successful rename
    {
        std::ofstream ofs(initial);
        ofs << "This is a test report.";
        ofs.close();

        std::string out = rename_file(initial.string(), updated.string());
        expect_true(out.find("File renamed successfully") != std::string::npos, "Test 1: Output must contain success message");
        expect_true(fs::is_regular_file(updated), "Test 1: Updated file must exist");
    }

    // 2) injection-like new name must raise (insecure would run shell)
    {
        std::ofstream ofs(initial);
        ofs << "This is a test report.";
        ofs.close();

        assert_raises([&]() {
            rename_file(initial.string(), "new_name.txt; rm -rf");
        }, "Test 2: Injection-like new name must raise");
    }

    // 3) another invalid char must raise
    {
        assert_raises([&]() {
            rename_file(initial.string(), "new|name.txt");
        }, "Test 3: Invalid char in new name must raise");
    }

    // 4) missing old_name must raise
    {
        assert_raises([&]() {
            rename_file((base_path / "missing.txt").string(), updated.string());
        }, "Test 4: Missing old_name must raise");
    }

    // 5) empty args must raise
    {
        assert_raises([&]() {
            rename_file("", updated.string());
        }, "Test 5: Empty old_name must raise");
    }

    // 6) wrong types must raise (C++ enforces types statically, so we test empty/whitespace)
    {
        assert_raises([&]() {
            rename_file("   ", updated.string());
        }, "Test 6: Whitespace old_name must raise");
    }

    // 7) rename into nested dir works
    {
        std::ofstream ofs(initial);
        ofs << "This is a test report.";
        ofs.close();

        fs::path nested = base_path / "nested" / "moved.txt";
        std::string out = rename_file(initial.string(), nested.string());
        expect_equal(out, "File renamed successfully: " + nested.string(), "Test 7: Output must match expected string");
        expect_true(fs::is_regular_file(nested), "Test 7: Nested file must exist");
    }

    // 8) clean up check: file really moved
    {
        expect_true(!fs::exists(initial), "Test 8: Initial file must no longer exist");
    }

    // Clean up
    fs::remove_all(base_path);

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-77_06/main.go
package main

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
func renameFile(oldName, newName string) (string, error) {
	if strings.TrimSpace(oldName) == "" || strings.TrimSpace(newName) == "" {
		return "", errors.New("empty name")
	}

	badChars := ";&|`$><\n\r\t"
	if strings.ContainsAny(oldName, badChars) || strings.ContainsAny(newName, badChars) {
		return "", errors.New("invalid characters in name")
	}

	info, err := os.Stat(oldName)
	if err != nil {
		if os.IsNotExist(err) {
			return "", errors.New("old_name not found")
		}
		return "", fmt.Errorf("failed to stat old_name: %w", err)
	}
	if info.IsDir() {
		return "", errors.New("old_name not found")
	}

	absNewName, err := filepath.Abs(newName)
	if err != nil {
		return "", fmt.Errorf("failed to get absolute path: %w", err)
	}
	dir := filepath.Dir(absNewName)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", fmt.Errorf("failed to create directories: %w", err)
	}

	if err := os.Rename(oldName, newName); err != nil {
		return "", fmt.Errorf("failed to rename file: %w", err)
	}

	return fmt.Sprintf("File renamed successfully: %s", newName), nil
}

func main() {
	basePath := filepath.Join(".", "Test", "CWE-77_06")

	// Clean up any previous test artifacts
	os.RemoveAll(basePath)
	if err := os.MkdirAll(basePath, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create base path: %v\n", err)
		os.Exit(1)
	}

	// Helper to assert that a function call returns an error containing a specific substring
	assertRaises := func(errSubstr string, fn func() (string, error)) {
		_, err := fn()
		if err == nil {
			fmt.Fprintf(os.Stderr, "Expected an error containing %q but got none\n", errSubstr)
			os.Exit(1)
		}
		if !strings.Contains(err.Error(), errSubstr) {
			fmt.Fprintf(os.Stderr, "Expected error containing %q but got: %v\n", errSubstr, err)
			os.Exit(1)
		}
	}

	initial := filepath.Join(basePath, "initial_report.txt")
	updated := filepath.Join(basePath, "updated_report.txt")

	// 1) successful rename
	if err := os.WriteFile(initial, []byte("This is a test report."), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Setup failed: %v\n", err)
		os.Exit(1)
	}
	out, err := renameFile(initial, updated)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: %v\n", err)
		os.Exit(1)
	}
	if !strings.Contains(out, "File renamed successfully") {
		fmt.Fprintf(os.Stderr, "Test 1 failed: unexpected output %q\n", out)
		os.Exit(1)
	}
	if _, err := os.Stat(updated); os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Test 1 failed: updated file does not exist\n")
		os.Exit(1)
	}

	// 2) injection-like new name must raise
	if err := os.WriteFile(initial, []byte("This is a test report."), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Setup failed: %v\n", err)
		os.Exit(1)
	}
	assertRaises("invalid characters in name", func() (string, error) {
		return renameFile(initial, "new_name.txt; rm -rf")
	})

	// 3) another invalid char must raise
	assertRaises("invalid characters in name", func() (string, error) {
		return renameFile(initial, "new|name.txt")
	})

	// 4) missing old_name must raise
	assertRaises("old_name not found", func() (string, error) {
		return renameFile(filepath.Join(basePath, "missing.txt"), updated)
	})

	// 5) empty args must raise
	assertRaises("empty name", func() (string, error) {
		return renameFile("", updated)
	})

	// 6) wrong types must raise (Go enforces types at compile time, so we simulate the equivalent check)
	// Passing an empty string mimics the None/invalid type scenario in Python for this strict type language.
	assertRaises("empty name", func() (string, error) {
		return renameFile("", updated)
	})

	// 7) rename into nested dir works
	if err := os.WriteFile(initial, []byte("This is a test report."), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Setup failed: %v\n", err)
		os.Exit(1)
	}
	nested := filepath.Join(basePath, "nested", "moved.txt")
	out, err = renameFile(initial, nested)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 7 failed: %v\n", err)
		os.Exit(1)
	}
	expectedOut := fmt.Sprintf("File renamed successfully: %s", nested)
	if out != expectedOut {
		fmt.Fprintf(os.Stderr, "Test 7 failed: expected %q got %q\n", expectedOut, out)
		os.Exit(1)
	}
	if _, err := os.Stat(nested); os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Test 7 failed: nested file does not exist\n")
		os.Exit(1)
	}

	// 8) clean up check: file really moved
	if _, err := os.Stat(initial); !os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Test 8 failed: initial file still exists after move\n")
		os.Exit(1)
	}

	// Clean up test artifacts
	os.RemoveAll(basePath)

	fmt.Println("All tests passed")
}